### PR TITLE
Hitissum/issue1

### PR DESCRIPTION
**Pull request by Hitissum**
Longest part of getting here was setting up the repository on my machine because I'm doing this from linux which I only started using recently as my daily driver as a sort of self challenge :)

Fixing the message order issue was trivial I took a look at the conversation endpoint that the backend console and flipped the desc to asc

Making the messages show up on the fly was a lot less simple, at first I noticed there was a missing async which was easy. Then I went on a tagent trying to solve it but had to go back to the beginning and reapproach the problem I realized that when I cleared the search bar the chat showed up so I decided to look into why that was the case and simplified the sendMessage method and axed postMessage because it wasn't used elsewhere and I could easily make it do both.

### CODE REVIEW

File: client/src/components/Home.js [hunk 1]
```diff
@@ -62,59 +62,33 @@ const Home = ({ user, logout }) => {
     });
   };
 
-  const postMessage = (body) => {
+  const postMessage = async (body) => {
```
Comment (temesghentekeste) [inline client/src/components/Home.js:65]: :+1:

File: client/src/components/Home.js [hunk 2]
```diff
@@ -62,59 +62,33 @@ const Home = ({ user, logout }) => {
...
     try {
-      const data = saveMessage(body);
-
-      if (!body.conversationId) {
-        addNewConvo(body.recipientId, data.message);
-      } else {
-        addMessageToConversation(data);
-      }
```
Comment (temesghentekeste) [inline client/src/components/Home.js:73]: This change is not required as the only fix in this code block is the async/await bug. Please undo the changes.

File: client/src/components/Home.js [hunk 3]
```diff
@@ -62,59 +62,33 @@ const Home = ({ user, logout }) => {
...
-
+      const data = await saveMessage(body);
```
Comment (temesghentekeste) [inline client/src/components/Home.js:67]: :+1:

File: client/src/components/Home.js [hunk 4]
```diff
@@ -62,59 +62,33 @@ const Home = ({ user, logout }) => {
...
-          convo.messages.push(message);
-          convo.latestMessageText = message.text;
-          convo.id = message.conversationId;
-        }
-      });
-      setConversations(conversations);
-    },
-    [setConversations, conversations]
-  );
-
```
Comment (temesghentekeste) [inline client/src/components/Home.js:94]: Same as above(undo the changes)

File: client/src/components/Home.js [hunk 5]
```diff
@@ -62,59 +62,33 @@ const Home = ({ user, logout }) => {
...
+        prev.map((convo) => {
+          if(convo.otherUser.id === recipientId){
+            const convoCopy = { ...convo };
+            convoCopy.messages.push(message);
+            convoCopy.latestMessageText = message.text;
+            convoCopy.id = message.conversationId;
+            return convoCopy;
+          } else {
+            return convo;
+          }
```
Comment (temesghentekeste) [inline client/src/components/Home.js:87]: You are on the right track, and :+1: for creating a shallow copy of the top-level state(Line#80). However, the state at the nested level(Line#81) is directly mutated. There are some hints given below to help you address the bug.

1. Observe how the state is copied in such methods as `addOnlineUser()` within the codebase. 
2. You can further reference this [document ](https://reactjs.org/docs/hooks-reference.html#:~:text=If%20your%20update%20function%20returns%20the%20exact%20same%20value%20as%20the%20current%20state%2C%20the%20subsequent%20rerender%20will%20be%20skipped%20completely.)on how to properly handle state immutability in React. 
3. The fix should also be done in the `addNewConvo()` method after you undo the changes.

Apply the above hints to address the mutating state bug?  :slightly_smiling_face:.

File: server/routes/api/conversations.js [hunk 6]
```diff
@@ -19,9 +19,9 @@ router.get("/", async (req, res, next) => {
         },
       },
       attributes: ["id"],
-      order: [[Message, "createdAt", "DESC"]],
+      order: [[Message, "createdAt", "ASC"]],
       include: [
-        { model: Message, order: ["createdAt", "DESC"] },
+        { model: Message, order: ["createdAt", "ASC"] },
```
Comment (temesghentekeste) [inline server/routes/api/conversations.js:24]: :100: 
But now, the conversation order is broken. The fix can be done either in the front end or backend. Backend implementation is recommended to preserve the current solution. :slightly_smiling_face:

File: server/routes/api/conversations.js [hunk 7]
```diff
@@ -68,7 +68,7 @@ router.get("/", async (req, res, next) => {
       }
 
       // set properties for notification count and latest message preview
-      convoJSON.latestMessageText = convoJSON.messages[0].text;
+      convoJSON.latestMessageText = convoJSON.messages[convoJSON.messages.length - 1].text;
```
Comment (temesghentekeste) [inline server/routes/api/conversations.js:71]: :100:

**Review by temesghentekeste — COMMENTED** [review]
Hi @Hitissum :wave:,

Thank you for your effort in this PR.

There are some suggestions given below that you need to tackle; when you address them please remember to ping us using` /hatchy pr` for a new review on the slack channel.

You can reply to the conversations using text comments or emojis but don't resolve them :pray:.

File: server/routes/api/conversations.js (hunk 6 above)
Comment (Hitissum) [reply server/routes/api/conversations.js:24]: I am unsure as to what you mean because the message order in the conversation appears correctly with this change

**Comment by Hitissum** [comment]
Whilst I was rather fond of my solution for removing redundant code and handling it more elegantly, I have reverted it and done the recommended changes instead. Thank you for your feedback

File: client/src/components/Home.js (hunk 4 above)
Comment (Hitissum) [reply client/src/components/Home.js]: The reasoning behind the refactor was to remove a lot of the redundancy of having two functions that do relatively the same thing (adding the message to the relevant conversation, where one assigns the conversation id) and consolidating them into a function that does both.

Comment (Hitissum) [reply client/src/components/Home.js]: I have made the changes, I am just curious as to the reasoning

File: client/src/components/Home.js [hunk 8]
```diff
@@ -62,14 +62,14 @@ const Home = ({ user, logout }) => {
...
+  const postMessage = async (body) => {
     try {
-      const data = saveMessage(body);
+      const data = await saveMessage(body);
 
       if (!body.conversationId) {
         addNewConvo(body.recipientId, data.message);
       } else {
-        addMessageToConversation(data);
+        addMessageToConversation(data.message);
```
Comment (uimarshall) [inline client/src/components/Home.js:72]: This change is not necessary. Please revert it 🙂.

File: client/src/components/Home.js [hunk 9]
```diff
@@ -80,41 +80,39 @@ const Home = ({ user, logout }) => {
...
+            const convoCopy = { ...convo };
+            convoCopy.messages = [ ...convoCopy.messages, message ];
+            convoCopy.latestMessageText = message.text;
+            convoCopy.id = message.conversationId;
+            return convoCopy;
+          } else {
+            return convo;
+          }
+        })
+      );
```
Comment (uimarshall) [inline client/src/components/Home.js:95]: 💯

File: client/src/components/Home.js [hunk 10]
```diff
@@ -80,41 +80,39 @@ const Home = ({ user, logout }) => {
...
+          if(convo.id === message.conversationId) {
+            const convoCopy = { ...convo };
+            convoCopy.messages = [ ...convoCopy.messages, message ];
+            convoCopy.latestMessageText = message.text;
+            return convoCopy;
+          } else {
+            return convo;
+          }
+        })
+      )
```
Comment (uimarshall) [inline client/src/components/Home.js:113]: 💯

File: server/routes/api/conversations.js [hunk 11]
```diff
@@ -59,16 +59,17 @@ router.get("/", async (req, res, next) => {
...
       if (onlineUsers.includes(convoJSON.otherUser.id)) {
         convoJSON.otherUser.online = true;
       } else {
         convoJSON.otherUser.online = false;
       }
-
+      
       // set properties for notification count and latest message preview
       convoJSON.latestMessageText = convoJSON.messages[0].text;
+      convoJSON.messages = convoJSON.messages.reverse();
```
Comment (uimarshall) [inline server/routes/api/conversations.js:72]: 💯

**Review by uimarshall — APPROVED** [review]
Thank you for addressing the comments from the previous review.
Consider this PR approved and feel free to merge after addressing the comments below! 👍
A small request: please don’t resolve the conversations, but feel free to reply to them.
